src/validation.cpp

```diff
@@ -2953,7 +2953,7 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
     do {
         boost::this_thread::interruption_point();
 
-        if (GetMainSignals().CallbacksPending() > 10) {
+        if (fSyncQueue && GetMainSignals().CallbacksPending() > 10) {
```

Comment:
Also codewise it looks overall good to me. Though i wonder if it now even makes any sense to skip the queue sync if `ActivateBestChain` gets called by CL/IS with CL having its own scheduler? Wasn't the issue that CL put `ActivateBestChain` (over edges) in the validation interface scheduler's queue and that basically blocked itself by waiting for the queue it runs from to be done with all tasks? 

@UdjinM6 please correct me if im wrong and explain me your full understanding of the deadlock as i was investigating a lot here how exactly it happens and i just want to fully get it 🙂

Reply:
Thee are two different cases:
1.
a) CL: LOCK->SyncWithValidationInterfaceQueue
b) PeerLogicValidation: (dead)LOCK

(b) (which is executed as a part of the queue flushing) is waiting for (a) to release the lock which will never happen because (a) is waiting for (b). Even if (a) would be in another scheduler it wouldn't matter, the lock would be held anyway because queue processing is blocked. Skipping SyncWithValidationInterfaceQueue lets (a) to finish and release the lock.

2.
a) PeerLogicValidation: LOCK->SyncWithValidationInterfaceQueue
b) CL: (dead)LOCK

In this case (b) (executed as a regular scheduler job) blocks _the scheduler itself_ not just the queue, it simply can't process anything at this point. Using another scheduler for (b) fixes this because now (b) blocks some another thread and the original scheduler can process all the tasks and release the lock which in its turn lets (b) to continue.

Reply:
Hmm.. soo.. if i look at the two cases you commented in #3476 it looks for me like the queue skipping just fixed one deadlock and introduced a new one. 

Lets assume we have: 

- Thread 1 > `A` > CConnman::ThreadMessageHandler 
- Thread 2 > `B` > Validation Interface Scheduler Thread
- Thread 3 > `C` > New ChainLocks Scheduler Thread
- Thread 4 > `D` > CInstantSendManager::WorkThreadMain
- Thread 5 > `E` > CSigSharesManager::WorkThreadMain
### Case 1 

**1a)** `B` runs `ActivateBestChain` it locks `cs_activateBestChain` and then since there are more than 10 tasks in its queue it adds another one and waits for this other one to finish which can never happen because it waits for itself. 

**1b)** Now `A` gets a new block message, wants to process it and therefor also calls `ActivateBestChain` and waits for `cs_activateBestChain` to release which won’t happen because of the above.. 

### Case 2 / Skipping the sync for CL code

If we skip the sync now Case 1 can’t longer happen because in **1a)** `B` will lock `cs_activateBestChain`, skips the sync (and with this it skips the wait on yourself), runs the rest of `ActivateBestChain` and then it releases `cs_activateBestChain` which means `A` will not wait forever for `cs_activateBestChain`.. This gives us just the new deadlock below:

**2a)** `A` gets a new block, wants to process it and calls `ActivateBestChain`. It locks `cs_activateBestChain` and since there are more than 10 tasks in the queue of `B` it adds another task to the queue of `B` and waits for this other task to finish.

**2b)** `B` which has at this point more than 10 tasks (one of them is `ActivateBestChain` which it got from the CL code with sync skipping) processes its tasks until it comes to the call of `ActivateBestChain` where it then waits for `cs_activateBestChain` to release whihc won't happen because `A` in **2a)** locked it. 

### Fix the deadlock / Adding Thread C

Soo.. the actual reason for the deadlock is IMO only calling `ActivateBestChain` from `B`. Means just adding `C` should be the fix for the initial deadlock and then deadlock 2 will never be introduced. 

And this is a problem now because we have the “Let Thread B” catch up thing here introduced in  #3376 https://github.com/dashpay/dash/blob/0662f170f09f5d1304d80e41754b59c9d56528d8/src/validation.cpp#L2956-L2961 
which leads to `X` waiting for `B` if `X` calls `ActivateBestChain` **and** there are more than 10 queue entries (which is why the deadlock happened randomly and more often on the mainnet it seems) and therefor `B` waiting for itself if `ActivateBestChain` gets called by `B` .  

So i would say having this "skip sync" code included makes no sense anymore because if `ActivateBestChain` gets called from instantsend code its called by `D` or `E` and if its called by CL its called by `C`. So there should be no more deadlock, not?

Does this make sense? Note: I'm tired now so i hope i didn't miss any other crazy magic 🤣

Reply:
I think I understand now, thanks 👍 Reverted, let's see :)